0028-web-monetization/0028-web-monetization.md

```diff
@@ -1,79 +1,296 @@
+Web Monetization makes use of [ILP/STREAM](../0029-stream/0029-stream.md) provide a high-level way to send money and data, while still providing tremendous flexibility.
 
 ## Flow
 
-- The user visits a webpage. The page is loaded and the page's javascript is run.
-- The site checks whether `window.monetize` is defined. This function is injected into the page by the browser or browser extension which provides Web Monetization.
-- If `window.monetize` is not defined, the webpage may show the user a message asking them to install a Web Monetization extension in order to support this site.
-- If `window.monetize` is defined, the webpage calls `monetize({ receiver })`, where `receiver` is their payment pointer. This function returns a promise.
-- The user's browser or Web Monetization extension decides whether to pay the page.
-- If the user's browser decides not to pay, an error is thrown from the `monetize` promise.
-- If the user's browser pays, the `monetize` promise resolves.
-- The webpage may run some code in order to thank the user or offer them additional content. This does not guarantee that the site was paid, because it runs client-side. If confirmation of receipt is required, the backend which provides the SPSP server can be queried.
+Web Monetization works in two stages: first, the user registers their provider with their browser. After the registration has been completed, the user can go to websites and use Web Monetization.
+
+### Registration
+
+- The user visits their provider's webpage.
+- The provider's webpage calls `window.WebMonetization.register`.
+- The browser will display a native pop-up box asking the user whether they want to set the handler proposed by the page.
+  - If the user confirms, `handlerUrl` is stored in the browser's state and `window.WebMonetization.register` resolves to true.
+  - If the user cancels the registration, `window.WebMonetization.register` will resolve to false.
+
+### Monetization
+
+- The user visits a webpage.
+- When the page wants to open an ILP/STREAM connection, it calls `window.WebMonetization.monetize()` with a `destinationAccount` and `sharedSecret`.
+  - The browser creates an iframe to the handler URL, reading from the state that was stored during [registration](#registration).
+  - An ILP/STREAM connection object is returned from the function for the site to use.
+- When the page wants to use the ILP/STREAM connection, they use the javascript STREAM API to send money and/or data. The browser sends outgoing ILP packets to the handler iframe using `postMessage`. The handler iframe forwards incoming packets by calling `window.parent.postMessage`.
 
 ## Specification
 
-#### monetize
+### Browser API / Polyfill API
+
+#### Register
 
-`window.monetize({ receiver: string }) -> Promise<void>`
+```ts
+window.WebMonetization.register({
```

Comment:
Based on the work done on the payment handler API I would recommend that this is done using [`Workers`](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API). Payment Handler is the only standard I am aware of that has a pattern of:
 1. Registration of service from Origin A
 2. Invocation from website at Origin B where browser mediates the communication from website to service

The pattern for extending the [`ServiceWorkerGlobalScope`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope) is to add the methods, events, and properties you need and then extend [`ServiceWorkerGlobalScope.registration`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration) with the necessary methods/properties that will be called during registration.

In the case of monetization there is no need to provide service meta-data other than the name so maybe we can skip the extra ServiceWorkerRegistration hooks for now and just identify the provider by origin.

This might look like this in WebIDL:

```webidl

partial interface ServiceWorkerGlobalScope {
  attribute EventHandler onmonetize;
};

[Constructor(DOMString type, MonetizeEventInit eventInitDict), Exposed=ServiceWorker]
interface MonetizeEvent : ExtendableEvent {
  readonly attribute USVString origin;
  readonly attribute DOMString destinationAccount;
  readonly attribute DOMString sharedSecret;
  void respondWith(Promise<IlpConnectionInit> handlerResponsePromise);
};

dictionary MonetizeEventInit : ExtendableEventInit {
  readonly attribute USVString origin;
  readonly attribute DOMString destinationAccount;
  readonly attribute DOMString sharedSecret;
};

dictionary IlpConnectionInit {
  readonly attribute DOMString connectionTag;
  readonly attribute double lastPacketExchangeRate;
  readonly attribute double minimumAcceptableExchangeRate;
  readonly attribute DOMString totalDelivered;
  readonly attribute DOMString totalSent;
  readonly attribute DOMString totalReceived;
}
```

It would be used by a provider like this:

```js
self.addEventListener('monetize', monetizeEvent => {
  console.log('Provider got monetize request', monetizeEvent)

  // 1. Peform work to create uplink (open WS to provider)
  // 2. Get handle on channel to send ilp packets back to website
  // 3. Return connection meta-data to be used to construct connection object
  const channel = monetizeEvent.ilpChannel
  const uplink = createUplink()

  uplink.on('data', reqFromUplink => {
    reqFromUplink.replyWith(channel.sendRequest(reqFromUplink))
  })

  channel.on('data', reqFromWebsite => {
    reqFromWebsite.replyWith(uplink.sendRequest(reqFromWebsite))
  })

  monetizeEvent.replyWith({
    connectionTag,
    lastPacketExchangeRate,
    minimumAcceptableExchangeRate,
    totalDelivered,
    totalSent,
    totalReceived	
  })
})
```

Reply:
Is the idea that the service worker running on the provider's origin would be pulling in the STREAM library? Then we would need a more complex RPC protocol in order to allow any ILP/STREAM method call to be triggered on the service worker from a page that uses web monetization

Reply:
No, it will only need to establish the BTP uplink to it's own server and then pass packets between that link and the website (via postMessage)